### ISSUE TITLE
Enforce the Liskov Substitution Principle for non-methods

### ISSUE THREAD

**Issue by AlexWaygood** [astral-sh/ty#2158]
We have implemented a basic version of the Liskov Substitution Principle to cover cases where methods are incompatibly overridden in subclasses. We also need to implement separate diagnostics to cover properties incompatibly overridden in subclasses, and mutable attributes incompatibly overridden in subclasses.

The reason why these are planned as separate diagnostics is because a large amount of code in the ecosystem unsoundly overrides mutable attributes covariantly, e.g.

```py
class A:
    x: int

class B(A):
    x: bool
```

Partly this is because mypy allowed this for years -- and still does, unless users explicitly opt into the `mutable-override` error code. Implementing these as separate diagnostics to our existing `invalid-method-override` diagnostic will therefore allow users to switch the mutable-attribute-override rule off specifically if it causes a large number of diagnostics on their code.

---

Sub-tasks (many of these may share a common implementation):

- [ ] Enforce Liskov for property types: this should cause us to emit a diagnostic:
    ```py
    class A:
        @property
        def f(self) -> int:
            return 42

    class B(A):
        @property
        def f(self) -> str:  ❌ Superclass returns `int`, subclass returns `str`, `str` is not a subtype of `int`
            return "42"
    ```

- [ ] Enforce that a writable attribute cannot be overridden with a read-only property:
    ```py
    class A:
        x: int

    class B(A):
        @property
        def x(self) -> int:  ❌ Superclass attribute is writable, subclass attribute is read-only
            return 42
    ```

    and

    ```py
    class A:
        @property
        def x(self) -> int:
            return 42

        @x.setter
        def x(self, value: int) -> None: ...

    class B(A):
        @property
        def x(self) -> int:  ❌ Superclass attribute is writable, subclass attribute is read-only
            return 42
    ```

- [ ] Enforce Liskov for attribute types:
    ```py
    class A:
        x: int

    class B(A):
        x: bool  ❌ Type of `x` attribute is invariant because it is mutable
    ```

- [ ] Enforce that a non-`Final` attribute cannot be overridden with a `Final` one
    ```py
    from typing import Final

    class A:
        x: int

    class B(A):
        x: Final[int]  ❌ Superclass attribute is writable, subclass attribute is read-only
    ```

- [ ] Enforce that a non-`ClassVar` attribute cannot be overridden with a `ClassVar` attribute:
    ```py
    from typing import ClassVar

    class A:
        x: int

    class B(A):
        x: ClassVar[int]  ❌ Superclass attribute is writable on instances, subclass attribute is not
    ```

- [ ] Enforce that a `ClassVar` attribute cannot be overridden with an implicit instance attribute, since `ClassVar`s can be mutated on the class object itself but implicit instance attributes cannot:
    ```py
    from typing import ClassVar

    class A:
        x: ClassVar[int]

    class B(A):
        def __init__(self):
            self.x: int  ❌ Superclass attribute is writable on the class object, subclass attribute is not
    ```

- [ ] Enforce that a non-`ReadOnly` `TypedDict` field cannot be overridden with a `ReadOnly` `TypedDict` field:
    ```py
    from typing import TypedDict, ReadOnly

    class A(TypedDict):
        x: int

    class B(A):
        x: ReadOnly[int]  ❌ Superclass field is writable, subclass attribute is read-only
    ```

**Comment by beauxq** [comment]
I think it's an important checklist item to check multiple levels of subclasses, not just one base class and one subclass -
because mypy and pyright each get things wrong in different places here:

```python
from typing import ClassVar, reveal_type

class B:
    x: ClassVar[int | str] = 0

class C(B):
    x = "random"

class D(C):
    x = 1  # mypy complains when it shouldn't

class E(D):
    x = b"x"  # pyright doesn't complain when it should


def foo(b: type[B]) -> None:
    x = b.x
    reveal_type(x)
    if isinstance(x, str):
        print(x + "y")
    else:
        reveal_type(x)
        print(x + 3)


foo(E)
```

**Comment by BobbyL2k** [comment]
I'm not sure if this example would be covered by this issue, or should I open another issue. I'm adding this here in hopes that if it is covered by this issue, it would useful as a snippet for testing.

```python
from abc import ABC


class Base:
    class Config(ABC):
        port: int


class Server(Base):
    # Should pass, since Server.Config extends Base.Config
    class Config(Base.Config):
        pass


class BadServer(Base):
    # Should error because it doesn't extends Base.Config
    class Config:
        pass
```

[Ty playground](https://play.ty.dev/8d9bda2e-844f-44b3-a908-8f56e8bd39d9)
[Pyright playground](https://pyright-play.net/?code=GYJw9gtgBAhgRgYygSwgBzCALlAggIQGEAoUhAGxgGcqp9qBTALmKjagutsLADthkAcwAUBQgEoW7aVAzYmKXllLFONKAGUGIAG7bh9Kg0mt2AYk0ALMAFdyAE1lcANFCrJeCBpu16QAOh5%2BISgGAA8sBl57WkMGQL4BQVM2NW5EoQNGBODBExl2NC4VNLoYey1dfTj88ytbB1CQcBAoOAYEGBsjFBx7MAYqXgByHHDI6NjsoKSUjkp1GaEpAqcaYiA)

**Comment by carljm** [comment]
#4035 discusses a use case we'll need to think about in terms of whether a descriptor that takes and returns `str` is a valid override for a simple `: str` attribute. It is fully substitutable for instance access but not for class access; that may need to be a soundness gap we accept, since it's practically useful to be able to replace a simple attribute with a descriptor.
